### ISSUE TITLE
glb files with embedded resources are embedded as data URI

### ISSUE THREAD

**Issue by dov** [syoyo/tinygltf#309]
**Describe the issue**

When embedding a buffer (and I assume an image) in a binary gltf file (glb), these resources are still encoded as data uri.  This is invalid according to the Khronos glTF validator, at https://github.khronos.org/glTF-Validator/ .

**To Reproduce**

1. Change the example build-gltf to save the resulting gltf file as follows:
```
  tinygltf::TinyGLTF gltf;
  gltf.WriteGltfSceneToFile(&m, "triangle.glb",
                           true, // embedImages
                           true, // embedBuffers
                           false, // pretty print
                           true); // write binary
```

2. Load the resulting `triangle.glb` file into https://github.khronos.org/glTF-Validator/ .

**Expected behaviour**

I would expect that if the `writeBinary` flag is on, then the binary resources would be embedded into the glb as binary blobs as per the gltf specification.

**Comment by syoyo** [comment]
Hmm. it reports valid on my side.

Changed build-gltf example as follows on Ubuntu:

```
  // Save it to a file
  tinygltf::TinyGLTF gltf;
  gltf.WriteGltfSceneToFile(&m, "triangle.glb",
                           true, // embedImages
                           true, // embedBuffers
                           false, // pretty print
                           true); // write binary
```

[triangle.glb.gz](https://github.com/syoyo/tinygltf/files/6067672/triangle.glb.gz)

![validator-result](https://user-images.githubusercontent.com/18676/109629731-7b0a3500-7b87-11eb-9dcc-70090b75fffb.png)

**Comment by dov** [comment]
Sorry! My mistake. I didn't realize that a glb file should only have one single buffer. I think it would make sense to throw an exception if trying to do `writeBinary` and `model->buffers.size() > 1`. But that's a different issues than the one I opened this bug about.

**Comment by syoyo** [comment]
@dov Although glb supports multiple chunks for embedded .bin data:

https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#chunks

Currently tinygltf always serialize all buffers into single binary chunk 

https://github.com/syoyo/tinygltf/blob/298c37a9549ae04bb991f0b2c77d270bccc66eb0/tiny_gltf.h#L7612

If you got an issue related with it, please open another issue with reproducible code/glTF data.

**Comment by dov** [comment]
> @dov Although glb supports multiple chunks for embedded .bin data:
> 
> https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#chunks

Actually, from the above link, you can have either 0 or 1 binary chunk, but not more.

> Currently tinygltf always serialize all buffers into single binary chunk
> 
> https://github.com/syoyo/tinygltf/blob/298c37a9549ae04bb991f0b2c77d270bccc66eb0/tiny_gltf.h#L7612

No, from the logic in the code, you promote the **first** buffer into a binary chunk, but keep the rest of the buffers URI data encoded. This was the cause of the gltf violation I got when I used more than binary buffer. 

> If you got an issue related with it, please open another issue with reproducible code/glTF data.

I'll do that. Thanks. :-)

**Comment by syoyo** [comment]
@dov Oh! Thanks for pointing out my mistake > Actually, from the above link, you can have either 0 or 1 binary chunk, but not more.

> No, from the logic in the code, you promote the first buffer into a binary chunk, but keep the rest of the buffers URI data encoded

I see.

> I'll do that. Thanks. :-)

Thanks!

**Comment by dbs4261** [comment]
Hi folks, I am not understanding what the resolution to this issue is. Ive been working on [extending the Open3D writer](https://github.com/dbs4261/Open3D/blob/model_writer/cpp/open3d/io/file_format/FileGLTF.cpp#L753) to be able to export textured GLTF files. I am testing this code by translating a textured OBJ file. The GLTF export works fine, however the GLB export does not seem to be working correctly. Even though the file passes validation with only a warning about the Data URI being used, 
some loaders, specifically the 3JS loader, cannot read the GLB file correctly. [BabylonJS](https://sandbox.babylonjs.com/) provides a useful error message `Unable to load from file:test.glb: /textures/0: Error while trying to load image: data:file:1677600676977/data:image/jpeg;base64,/9j/4AA`

So what is the correct method for writing a textured model to GLB with embedded images?

**The Khronos GLTF validator report:**
```
{
    "uri": "test.glb",
    "mimeType": "model/gltf-binary",
    "validatorVersion": "2.0.0-dev.3.8",
    "validatedAt": "2023-02-28T16:29:56.388Z",
    "issues": {
        "numErrors": 0,
        "numWarnings": 1,
        "numInfos": 1,
        "numHints": 0,
        "messages": [
            {
                "code": "URI_GLB",
                "message": "URI is used in GLB container.",
                "severity": 2,
                "pointer": "/images/0/uri"
            },
            {
                "code": "DATA_URI_GLB",
                "message": "Data URI is used in GLB container.",
                "severity": 1,
                "pointer": "/images/0/uri"
            }
        ],
        "truncated": false
    },
    "info": {
        "version": "2.0",
        "generator": "Open3d",
        "resources": [
            {
                "pointer": "/buffers/0",
                "mimeType": "application/gltf-buffer",
                "storage": "glb",
                "byteLength": 107995840
            },
            {
                "pointer": "/images/0",
                "mimeType": "image/jpeg",
                "storage": "data-uri",
                "image": {
                    "width": 8192,
                    "height": 8192,
                    "format": "rgb",
                    "bits": 8
                }
            }
        ],
        "animationCount": 0,
        "materialCount": 1,
        "hasMorphTargets": false,
        "hasSkins": false,
        "hasTextures": true,
        "hasDefaultScene": true,
        "drawCallCount": 1,
        "totalVertexCount": 2999870,
        "totalTriangleCount": 1000000,
        "maxUVs": 1,
        "maxInfluences": 0,
        "maxAttributes": 3
    }
}
```

**Comment by syoyo** [comment]
@dbs4261

It looks like embedded resources should be all point to buffer data through `bufferView` and data URI(`uri`) is not supported for glb: 

https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#images

See this example

https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/BoxTextured/glTF-Binary

```
"images":[{"bufferView":3,"mimeType":"image/png"}]
```

**Comment by dbs4261** [comment]
Ok, I understand where you're going with that. How does the image data get added to the buffer? Do I need to do it manually or will tinygltf add it for me? I was unsure how to use the bufferview parameter as I wasnt sure how putting the image into the png/jpg container would work.

**Comment by TheExDeus** [comment]
Hi! In my case I only have mesh data (vertices, indices) without any other binary data. And if I export as GLB with embedBuffers=True to make the GLB self-contained all validators show warnings about `URI is used in GLB container.`. How to get around that? It is because of the issue mentioned above (`No, from the logic in the code, you promote the first buffer into a binary chunk, but keep the rest of the buffers URI data encoded.`) or some other reason?

**Comment by ClemensX** [comment]
I came across this page via google, while I encountered similar problems. Maybe this info still has worth for somebody:

- Indeed, if you want to create a glm put all images into the one large buffer and set mime type correctly.
- This may sound obvious, but it still took me some time to figure out: Setting mime type and copying image data to the buffer means that the byte stream must be in the right format! I tried to push uncompressed image data into the buffer, like the ones we get from using stbi_load. Instead, if you have uncompressed image data, write it to the buffer with e.g. stbi_write_png_to_mem().

If you want to look at code you might be interested in looking up my glb exporter for WorldCreator landscapes: https://github.com/ClemensX/WorldCreatorToGLTF
